Feature/Programming-Exercise/Programming submission overview + trigger all (failed)

thilo-behnke opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested the changes and all related features on the test server https://artemistest.ase.in.tum.de.
- [x] I documented my source code using the JavaDoc / JSDoc style.
- [x] I added integration test cases for the server (Spring) related to the features
- [x] I added integration test cases for the client (Jest) related to the features
- [x] I added screenshots/screencast of my UI changes
- [x] I translated all the newly inserted strings

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->

Add a button to retrigger all/only failed student submissions of an exercise.
This is a good opportunity to also show some information in the participations list (e.g. how many submissions are running atm or how many have failed).

### Description
<!-- Describe your changes in detail -->

This PR introduces a submission state component for the instructor:
- Overview: How many latest submissions are running, how many have failed?
- Estimated time of arrival for results based on the number of currently running builds (2 minutes + 4 minutes for every 100 running builds)
- Trigger all: Will trigger **all** participations again (create submission from last commit hash with submissionType Instructor)
- Trigger failed: Will only trigger those participations that have failed on their last submission

Besides the implementation I also had to change the error handling for the websocket.service, as it would end the stream (subject.error makes the stream stop, which is a problem for ongoing streams for websocket subscriptions).

Todos:
- [x] Implementation
- [x] Refactoring and docs
- [x] Update existing client tests
- [x] Write client service tests
- [x] Write client integration tests
- [x] Write server integration tests 

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

1. Log in to Artemis
2. Navigate to Course Administration > Programming Exercise > Participations (#/exercise/797/participation has multiple participations)
3. Trigger builds, stop builds to make the submission fail, trigger failed, trigger all...

:warning: It can happen that your builds are blocked by Artemis CI, so it can happen that all submissions are considered failed, just because they are queued so long... We are looking at ways to improve the Bamboo config to resolve this issue.

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->

![image](https://user-images.githubusercontent.com/28230611/64113436-9aee4c00-cd8a-11e9-88c6-2dc036ae3de6.png)

jpbernius commented on src/main/webapp/app/programming-submission/programming-submission.service.ts:
```diff
@@ -291,12 +286,12 @@ export class ProgrammingSubmissionService implements IProgrammingSubmissionServi
-                map(submissions => {
-                    return Object.entries(submissions).map(([participationId, submission]) => [parseInt(participationId, 10), submission]);
-                }),
+                map(Object.entries),
```
I see that this approach was already used before, but could we also pass a domain object here to make it easier to understand?

thilo-behnke replied:
Looking at this again today, you are right. I refactored the code to always use the `ProgrammingSubmissionStateObj` type.

About the use of `Object.entries` here: It's not possible here for the server to already submit the domain object, as the server doesn't check the submissionDates of the latest submissions (if it's older than 2 minutes, we consider it failed) so it doesn't return the state of submission, just the latest submission if it has no result yet.

krusche commented:
The status in the upper right corner looks awesome!

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/dto/BuildTriggerDTO.java:
```diff
@@ -0,0 +1,19 @@
+/**
+ * DTO used for specificing participationIds of programming exercises for which the build should be triggered.
+ */
+public class BuildTriggerDTO {
```
Do we really need this, if it is just a wrapper for a List of Longs? Could we not just send a `List<Long>` object instead?
If we really want to keep this and someone else wants to reuse this: Should we then not rename this to ParticipationListDTO?

thilo-behnke replied:
I was not aware that it is possible to use a list/set here directly. I'll remove the DTO.

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingSubmissionResource.java:
```diff
@@ -191,11 +252,24 @@ public ProgrammingSubmissionResource(ProgrammingSubmissionService programmingSub
+     * @param submission ProgrammingSubmission that was just created.
+     */
+    private void notifyUserAndTriggerBuildForNewSubmission(ProgrammingSubmission submission) {
+        messagingTemplate.convertAndSend("/topic/participation/" + submission.getParticipation().getId() + Constants.PROGRAMMING_SUBMISSION_TOPIC, submission);
```
This looks a bit strange for me. Why do we first send the new submission to the user and then trigger the build on the CI server? Should it not be the other way round?

What happens if the CI server does not respond or throws an exception? See e.g. https://docs.atlassian.com/atlassian-bamboo/REST/5.5.0/#d2e823

thilo-behnke replied:
You are right, this is strange.
The handling of informing the user vs. triggering the build is inconsistent.
I adjusted it for most parts to send an error object to the user in case the communication to Bamboo fails, though there is a case in the service for `resumeExercise` that is not easy to handle without restructuring, so I added a TODO.

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingSubmissionResource.java:
```diff
@@ -191,11 +252,24 @@ public ProgrammingSubmissionResource(ProgrammingSubmissionService programmingSub
+    private void notifyUserTriggerBuildForNewSubmissions(Collection<ProgrammingSubmission> submissions) {
         for (ProgrammingSubmission submission : submissions) {
-            messagingTemplate.convertAndSend("/topic/participation/" + submission.getParticipation().getId() + "/newSubmission", submission);
+            notifyUserAndTriggerBuildForNewSubmission(submission);
```
Did you test this for a couple of hundred of students? Or even thousands?
Please talk to @sleiss 
He has a script to generate up to 500 participations with existing build results with the 500 artemis_test_users that we have. If we need more, we can create more test users

thilo-behnke replied:
I will discuss it with @sleiss, but I think we can merge this PR when it's done regardless as the feature is pretty isolated from the general submission process.

krusche replied:
Sounds good, I agree, we can still test this after the merge. Would just be interesting how the performance characteristics of the for loop are with many users.

krusche commented on src/main/webapp/app/entities/participation/participation.model.ts:
```diff
@@ -33,6 +33,7 @@ export abstract class Participation implements BaseEntity {
     public results: Result[];
     public submissions: Submission[];
     public type: ParticipationType;
+    public exercise: Exercise;
```
I am not sure if this is a good idea, because the subclasses of Participation in the client already have an attribute `public exercise: Exercise;` or `public programmingExercise: ProgrammingExercise;`

We should at least make sure that `SolutionProgrammingExerciseParticipation` and `TemplateProgrammingExerciseParticipation` don't have the exercise twice then.

Is it possible in TypeScript to refine the type of an attribute in the subclass? Then we could simply rename `public programmingExercise: ProgrammingExercise;` --> `public exercise: ProgrammingExercise;` in `SolutionProgrammingExerciseParticipation` and `TemplateProgrammingExerciseParticipation`.

What do you think?

thilo-behnke replied:
You are right, this would be inconsistent and incorrect for the Template/SolutionParticipation (which I did not test here as they are not relevant for this feature). I removed the attribute and made sure it is not used anymore.
I think the solution we have atm is correct and I didn't think about the change enough.

krusche commented:
Some additional questions:
1) What happens if the results is retrieved after 2 min? Is it then still updated and the failed count is decreased?
2) I think the participations list is a good candidate to dynamically adapt the 2 min constant. A very simple approach could e.g. be the following: per 100 submissions that are triggered, we increase the timeout by 1 additional minute. This means we would have 2min + 15 min in case 1500 submissions are triggered with the button click. Would this be possible or too complex because the timer is handled in a sub component? 
3) Why is the trigger all button grey? I would make it red and even add a popup asking for confirmation. This is not a button that should be clicked by accident. Please make sure that it is only displayed for instructors!!!
4) If not yet available, we should definitely have some additional information about what the buttons do in a small overlay, when the user hovers over them.

thilo-behnke commented:
> Some additional questions:
> 
> 1. What happens if the results is retrieved after 2 min? Is it then still updated and the failed count is decreased?
> 2. I think the participations list is a good candidate to dynamically adapt the 2 min constant. A very simple approach could e.g. be the following: per 100 submissions that are triggered, we increase the timeout by 1 additional minute. This means we would have 2min + 15 min in case 1500 submissions are triggered with the button click. Would this be possible or too complex because the timer is handled in a sub component?
> 3. Why is the trigger all button grey? I would make it red and even add a popup asking for confirmation. This is not a button that should be clicked by accident. Please make sure that it is only displayed for instructors!!!
> 4. If not yet available, we should definitely have some additional information about what the buttons do in a small overlay, when the user hovers over them.

**Current state:**
- The requested items are done
- :warning: I need to write some new tests and manually test it on the server

**Concerning the simulations:**
I talked to @sleiss, but think we should leave it out of this PR as the feature is only visible to instructors and doesn't change the general submission behavior from the code-editor, local git, etc.

**Your other points:**
1. Before the service would not react to it, but I changed it and also adjusted the client tests. Now as long as there is no newer submission, the client will react to the incoming result and then mark the pending submission as not pending.
2. I thought this sounded cool and integrated this into the client. The ETA for results is also visualized in the UI (see updated screenshot)
3. I also thought about this, you are right. I made the button red and added a modal that the user needs to confirm.
4. There are tooltips in the programmingExercise.json.

krusche commented on src/main/webapp/i18n/de/programmingExercise.json:
```diff
@@ -88,7 +88,14 @@
+            "resubmit": "Die letzte Abgabe erneut bewerten. Diese Aktion führt einen Build auf dem Repository der Teilnahme aus ohne einen Commit zu machen.",
+            "resubmitAll": "Alle erneut ausführen",
+            "resubmitAllTooltip": "Führe die letzte Abgabe aller Teilnahmen erneut aus. Das Ergebnis wird als bewertet gekennzeichnet.",
+            "resubmitAllDialog": "ACHTUNG: Die erneute Abgabe aller Teilnahmen ist eine sehr aufwändige Operation. Diese Aktion wird für jede Teilnahme dieser Übung einen CI Build starten!",
```
Please something like "Das bedeutet, dass alle Ergebnisse der Studierenden neu berechnet und überschrieben werden!"

thilo-behnke replied:
Done!

krusche commented on src/main/webapp/i18n/en/programmingExercise.json:
```diff
@@ -88,7 +88,14 @@
+            "resubmit": "Resubmit the last submission. This will execute a build run on the participation's repository without making a commit.",
+            "resubmitAll": "Trigger all",
+            "resubmitAllTooltip": "Trigger the last submission of all participations again. The results will be rated.",
+            "resubmitAllDialog": "WARNING: Triggering all participations again is a very expensive operation. This action will start a CI build for every participation in this exercise!",
```
Please something like "This means, that all results of the students will be calculated again and will be overridden!"

thilo-behnke replied:
Done!

thilo-behnke commented:
> > Some additional questions:
> > 
> > 1. What happens if the results is retrieved after 2 min? Is it then still updated and the failed count is decreased?
> > 2. I think the participations list is a good candidate to dynamically adapt the 2 min constant. A very simple approach could e.g. be the following: per 100 submissions that are triggered, we increase the timeout by 1 additional minute. This means we would have 2min + 15 min in case 1500 submissions are triggered with the button click. Would this be possible or too complex because the timer is handled in a sub component?
> > 3. Why is the trigger all button grey? I would make it red and even add a popup asking for confirmation. This is not a button that should be clicked by accident. Please make sure that it is only displayed for instructors!!!
> > 4. If not yet available, we should definitely have some additional information about what the buttons do in a small overlay, when the user hovers over them.
> 
> **Current state:**
> 
> * The requested items are done
> * warning I need to write some new tests and manually test it on the server
> 
> **Concerning the simulations:**
> I talked to @sleiss, but think we should leave it out of this PR as the feature is only visible to instructors and doesn't change the general submission behavior from the code-editor, local git, etc.
> 
> **Your other points:**
> 
> 1. Before the service would not react to it, but I changed it and also adjusted the client tests. Now as long as there is no newer submission, the client will react to the incoming result and then mark the pending submission as not pending.
> 2. I thought this sounded cool and integrated this into the client. The ETA for results is also visualized in the UI (see updated screenshot)
> 3. I also thought about this, you are right. I made the button red and added a modal that the user needs to confirm.
> 4. There are tooltips in the programmingExercise.json.

**Update for 2:**
I had to increase the time penalty to _4 minutes_ per 100 pending submissions, as this seemed to fit the build speed best (we might have to increase it further).

We could also integrate another category "overdue submissions" to better visualize the process.

ungaralex commented on src/main/webapp/app/entities/programming-exercise/actions/programmming-exercise-instructor-submission-state.component.ts:
```diff
@@ -0,0 +1,121 @@
+        }
+    }
+
+    triggerBuildOfAllSubmissions() {
```
Please add documentation

thilo-behnke replied:
You are right, also the naming of the method was bad.

ungaralex commented:
Reloading the page while building displays all running builds as failed.

ungaralex commented:
In general, I really like the new feedback on the currently running builds. 
But we should really, in a future PR, somehow block the "trigger all" button as long as there are still some builds running. Especially right after triggering all builds, an impatient user could immediately try to trigger all builds again, since he doesn't see the "running" status right after the first trigger. 
This would unnecessarily increase the load on Bamboo for huge courses

thilo-behnke commented on src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java:
```diff
@@ -197,4 +203,28 @@ public ProgrammingSubmission createSubmissionWithLastCommitHashForParticipation(
+    public List<ProgrammingSubmission> createSubmissionWithLastCommitHashForParticipationsOfExercise(List<ProgrammingExerciseStudentParticipation> participations,
+            SubmissionType submissionType) {
+        return participations.stream().map(participation -> {
+            ProgrammingSubmission submission;
```
You are right!

thilo-behnke commented on src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingSubmissionResource.java:
```diff
@@ -163,9 +167,67 @@ public ProgrammingSubmissionResource(ProgrammingSubmissionService programmingSub
+        if (programmingExercise == null) {
+            return notFound();
+        }
+        if (!authCheckService.isAtLeastInstructorForExercise(programmingExercise)) {
```
Yes, because the user could be an Instructor, but not for this exercise.

jpbernius commented on src/main/webapp/app/programming-submission/programming-submission.service.ts:
```diff
@@ -18,34 +17,49 @@ export enum ProgrammingSubmissionState {
+ * @param toBeDetermined either a ProgrammingSubmission or a ProgrammingSubmissionError.
+ */
+const checkIfSubmissionIsError = (toBeDetermined: ProgrammingSubmission | ProgrammingSubmissionError): toBeDetermined is ProgrammingSubmissionError => {
+    return !!(toBeDetermined as ProgrammingSubmissionError).error;
```
See question on similar Snippet above.

thilo-behnke replied:
See comment above.

thilo-behnke commented on src/main/java/de/tum/in/www1/artemis/service/connectors/BambooService.java:
```diff
@@ -226,6 +227,7 @@ public void triggerBuild(ProgrammingExerciseParticipation participation) {
                 Map.class);
         } catch (Exception e) {
```
You are right!

thilo-behnke commented on src/main/webapp/app/code-editor/service/code-editor-repository.service.ts:
```diff
@@ -20,6 +21,18 @@ export enum RepositoryError {
+ * @param toBeDetermined either a FileSubmission or a FileSubmissionError.
+ */
+const checkIfSubmissionIsError = (toBeDetermined: FileSubmission | FileSubmissionError): toBeDetermined is FileSubmissionError => {
+    return !!(toBeDetermined as FileSubmissionError).error;
```
If e.g. FileSubmission would also contain error, this function would not work as expected anymore :( But I think there is no better way to do this in TS: https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards

thilo-behnke commented on src/main/webapp/app/entities/programming-exercise/actions/programmming-exercise-instructor-submission-state.component.ts:
```diff
@@ -50,23 +58,64 @@ export class ProgrammmingExerciseInstructorSubmissionStateComponent implements O
+            <div class="modal-body">
+                <jhi-alert-error></jhi-alert-error>
+                <p jhiTranslate="artemisApp.programmingExercise.resubmitAllDialog">
+                    WARNING: Triggering all participations again is a very expensive operation. This action will start a CI build for every participation in this exercise!
```
Atm there is no normal alert that should be displayed in the modal, as it immediately closes when confirm is pressed. I left the alert-error inside but it could also be removed for the current implementation.

thilo-behnke commented:
This PR has some issues with the Result ETA, but we should merge it anyway:
- When the page is reloaded, the result ETA is reset to 2 minutes
- This makes previously building submissions with a higher ETA considered failed
- This then doesn't increase the ETA...

I will work on this on a new PR.
